Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+        }
+
+        NEGATIVE_FLAG((&left)) = 0;
+
+        if ((error = NSDecimalAdd(result, &left, rightOperand, roundingMode)) != NSCalculationNoError) {
+            return error;
+        }
+        // reset the negation flag.
+        NEGATIVE_FLAG(result) = 1;
+        return error;
+    }
+
+    if ((error = NSDecimalNormalize(&left, &right, roundingMode)) != NSCalculationNoError) {
+        return error;
+    }
+
+    NSComparisonResult comparison = NSDecimalCompare(leftOperand, rightOperand);
+    if (comparison == NSOrderedSame) {
+        SET_TO_ZERO(result);
+        return NSCalculationNoError;
+    }
+
+    if (IS_NEGATIVE(leftOperand) && IS_NEGATIVE(rightOperand)) {
+        NEGATIVE_FLAG((&left)) = 0;
+        NEGATIVE_FLAG((&right)) = 0;
+        if ((error = _evalSubtractSameNumbers(result, &right, &left, comparison)) != NSCalculationNoError) {
+            return error;
+        }
+    } else {
+        if ((error = _evalSubtractSameNumbers(result, &left, &right, comparison)) != NSCalculationNoError) {
+            return error;
+        }
+    }
+
+    NSDecimalCompact(result);
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCompact(NSDecimal* number) {
+    if ((number == NULL) || NSDecimalIsNotANumber(number)) {
+        return;
+    }
+
+    int indexer = 0;
+    // remove all the leading and trailing 0s of the mantissa.
+    while ((indexer < number->_length) && (MANTISSA(number)[indexer] == 0)) {
+        indexer++;
+    }
+
+    // shift the mantissa by indexer
+    if (indexer > 0) {
+        memmove(&MANTISSA(number)[0], &MANTISSA(number)[indexer], sizeof(unsigned short) * (LENGTH(number) - indexer));
+        LENGTH(number) -= indexer;
+    }
+
+    // remove trailing 0s
+    indexer = LENGTH(number) - 1;
+    while ((indexer >= 0) && (MANTISSA(number)[indexer] == 0) && (EXPONENT(number) < EXPONENT_MAX)) {
+        LENGTH(number)--;
+        EXPONENT(number)++;
+        indexer--;
+    }
+
+    if (LENGTH(number) == 0) {
+        SET_TO_ZERO(number);
+    }
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) || (rightOperand != NULL));
+
+    if (NEGATIVE_FLAG(leftOperand) != NEGATIVE_FLAG(rightOperand)) {
+        // miss match in sign.
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    int leftValue = EXPONENT(leftOperand) + LENGTH(leftOperand);
+    int rightValue = EXPONENT(rightOperand) + LENGTH(rightOperand);
+    if (leftValue > rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedAscending : NSOrderedDescending;
+    }
+    if (leftValue < rightValue) {
+        return IS_NEGATIVE(rightOperand) ? NSOrderedDescending : NSOrderedAscending;
+    }
+
+    // same length
```

Comment:
>same length [](start = 7, length = 11)

length is again confusing here, the effective length (matissa + exp) is same, but the length of mantissa may be different.